File: MTLab4Waiter/Philosopher.cs
namespace MTLab4Waiter;

public class Philosopher
{
	private Table _table;
    private readonly Waiter _waiter;
    private readonly int _philosopherId;
    private readonly int _leftFork;
    private readonly int _rightFork;

    public Philosopher(int id, Table table, Waiter waiter)
    {
        _philosopherId = id;
        _table = table;
        _waiter = waiter;
        _rightFork = id;
        _leftFork = (id + 1) % 5;
        var thread = new Thread(Run);
        thread.Start();
    }

    private void Run()
    {
        for (var i = 0; i < 10; i++)
        {
            Console.WriteLine($"Philosopher {_philosopherId} is thinking ({i + 1})");

            _waiter.RequestPermission();

            _table.TakeFork(_rightFork);
            _table.TakeFork(_leftFork);

            Console.WriteLine($"Philosopher {_philosopherId} is eating ({i + 1})");

            _table.ReleaseFork(_rightFork);
            _table.ReleaseFork(_leftFork);

            _waiter.ReleasePermission();
        }
    }
}

File: MTLab4JavaWaiter/src/Philosopher.java
public class Philosopher extends Thread {
    private final Table table;
    private final Waiter waiter;
    private final int leftForkIndex, rightForkIndex;
    private final int philosopherId;

    public Philosopher(int philosopherId, Table table, Waiter waiter) {
        this.philosopherId = philosopherId;
        this.table = table;
        this.waiter = waiter;

        rightForkIndex = philosopherId;
        leftForkIndex = (philosopherId + 1) % 5;

        start();
    }

    @Override
    public void run() {
        for (int iteration = 0; iteration < 10; iteration++) {
            System.out.println("Philosopher " + philosopherId + " is thinking (" + (iteration + 1) + " time)");

            waiter.requestPermission();

            table.getFork(rightForkIndex);
            table.getFork(leftForkIndex);

            System.out.println("Philosopher " + philosopherId + " is eating (" + (iteration + 1) + " time)");

            table.putFork(leftForkIndex);
            table.putFork(rightForkIndex);

            waiter.releasePermission();
        }
    }
}